File: 90-subsets-ii/90-subsets-ii.java
class Solution {
   List<List<Integer>> ans = new ArrayList<>();
    int[] nums;
    Stack<Integer> s = new Stack<>();

    void solve(int idx) {
        ans.add(new ArrayList<>(s));

        for (int i = idx; i < nums.length; i++) {
            if (idx < i && nums[i] == nums[i - 1])
                continue;
            s.push(nums[i]);
            solve(i + 1);
            s.pop();
        }
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        this.nums = nums;
        solve(0);
        return ans;
    }
}

File: 90-subsets-ii/90-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = [[]]
        prev = [[]]
        for i in range(len(nums)):
            tempans = []
            if i>0 and nums[i]==nums[i-1]:
                for subset in prev:
                    tempans.append(subset+[nums[i]])
            else:
                for subset in ans:
                    tempans.append(subset+[nums[i]])
            prev = tempans
            ans.extend(tempans)
        return ans
            